### ISSUE TITLE
Issue with default status

### ISSUE THREAD

**Issue by JohOply** [axemwa/NoStrip#12]
Hi there

When default status is off, then I use /ns toggle, it displays "enable" but is not (axe is still disabled)

So I need to toggle again, then it's works correctly until the next session

I use 2.0 version
https://i.imgur.com/KgGh3aU.png

**Comment by axemwa** [comment]
Did you delete your old jar and configs and add axe permission?

**Comment by JohOply** [comment]
Absolutely

I'm OP, with admin group ( '*' permission)

The fact is it works, but needs command executed twice (instead of one time like usual)

**Comment by axemwa** [comment]
Tried without being Op?

**Comment by axemwa** [comment]
I'll have a look either way

**Comment by axemwa** [comment]
Oh yeah I think I see the problem, will get onto it

**Comment by axemwa** [comment]
Does it work if you're not OP?

**Comment by axemwa** [comment]
Also what permissions do you have set currently?

**Comment by axemwa** [comment]
Seems to work as intended if not Opp'd but will try to fix it for Ops

**Comment by axemwa** [comment]
Also please mention any permissions other than .* that you may be using, are there any issues with negating permissions for ops?

**Comment by JohOply** [comment]
I didn't try with no op
No have negated permissions, only '*'

**Comment by axemwa** [comment]
So you tried this for example, negating permissions works fine for me?
/lp group default perm set nostrip.* false?

**Comment by axemwa** [comment]
Hmm all seems to work fine for me, did you try negate the permission like this?
/lp group default perm set nostrip.* false ? While having axe and toggle permissions set to how u want them?

**Comment by JohOply** [comment]
I just tried with player perms: nostrip.axe.* and nostrip.toggle
By default toggle off

I can't strip even if I use /ns toggle

EDIT: MY BAD

**Comment by axemwa** [comment]
did u also add nostrip.toggle to permission as well as the axe permission for the axe u wana use?

**Comment by JohOply** [comment]
Ok my bad for last comment, i was in a protected region

Your tips works, it's weird by the way... So it's needed to negate nostrip.* for default groups to works as intented.
If permission is not negate, there is the issue even if i'm a player (no OP perms)

**Comment by axemwa** [comment]
Yes because most people who own a server don't op their players, unless they admin, so it has to be setup differently for Ops, it also depends on the plugin and how they have coded their permissions but negating them for default or whatnot usually does the trick

**Comment by JohOply** [comment]
I think you didn't understand what I mean

In normal behaviour, because NoStrip has dedicated all perms to OP (like you said in plugin page), it's not needed to negate permissions for normal players.
In our case, it is needed to negate. 

If you want to keep this behavior, you need to warn people, because it's not obvious to have to negate permissions :)

**Comment by axemwa** [comment]
No I think you misunderstood maybe lost in translation, on the Plugin page I did indeed say Permissions need to be negated for Ops and that is the warning. I shouldn't have to warn people anyway if you're giving your players Op you should already know what does and doesn't need to be set in permissions as everybody has their server setup differently. 

This plugin was originally intended for default players not Opped players it just so happens that it's compatible with opped players if you negate permissions as I stated on the plugin page.
And like I said it depends on how the plugin has been coded not all plugins set permissions the same and most server owners already know that so warning them about it should not be necessary even though I did warn on the plugin page.

![image](https://user-images.githubusercontent.com/37608633/63224473-06b9ac00-c204-11e9-9221-8e83fe730e1c.png)

**Comment by JohOply** [comment]
Like I said, the issue appear even for non OP player, if permissions are not negate for default group. That's why i'm not sure to understand the link

Also, seems the first time I see permissions with this behavior. In usual cases, permissions are or aren't enable for players and / or OP. If enable for player, needs to be negate if don't want to give perms to players.

With NoStrip, perms are enable only for OP by default. So that's why it's weird the issue I spoke about in my first message appears, even for OP and Non-op players

**Comment by JohOply** [comment]
Anyway, with your trick it works. I just think you have to be more clear about this element in your topic (from my point of view it's not)

**Comment by JohOply** [comment]
Also, one more time, it's not about enable or disable perms for OP.
The point of this ticket is about a weird behaviour of the "ns toggle" command, which needs to be enable twice to works (only if perms are not negate for default groups)

**Comment by axemwa** [comment]
Non Op permissions work as they should there's no way I can be any clearer than I have descriptions and explanations are all in the plugin page, plus ive had people testing this for me and default non-opped players permissions and messages all work fine

**Comment by axemwa** [comment]
If u have an example feel free to say what I should write to make it clearer

**Comment by axemwa** [comment]
I don't have that weird enable message come up twice if permissions are set correctly whether I'm Op or not you are the only one who is experiencing this issue so it just means you haven't set something up correctly or don't understand how the plugin is supposed to work. Every plugin works differently and this is the only NoStrip plugin of it's kind so that's probably why you're not used to it cause it works differently to most plugins since it prevents stripping. I just had 3 more people test this problem ur having and when permissions are set correctly the enabled message doesn't appear twice whether they are op or not and they said I explained things clearly enough already I think you just mistranslated or weren't aware how this plugin handles everything since it's one of a kind so far which is ok everyone makes mistakes. I'm here to help either way if you have any suggestion as to any more info I could add to the plugin page to help let me know what exactly I should write I'm here to help.

**Comment by JohOply** [comment]
Until today, I never had to negate permissions which are already negated by the plugin itself. So that's why I don't understand the permission behaviour.

For instance, in the VentureChat plugin, I have to negate permissions because they are enabled by default to everyone and I want to deny them. In NoStrip, no any permission are enabled by default. That's why I told you it's not obvious to negate permissions to the default group.

The behaviour is understandable, but like I said it's not obvious because your plugin is the first one I experienced with this permissions behaviour (in 8 years and hundred of plugins!). That why in my point of view you have to tell clearly:
```You have to negate nostrip.* to default group (eg: LuckPerms Eg: /lp group default perm set worldedit.selection.* false). Then enable each permission to any group you want```
(english isn't my native language, but something like this could be clearer).

Everything I told you is from my point of view only, but if I misunderstood something or experienced this issue, it's possible I'm not alone. All my previous issue were real ;)
It just needs to understand how your plugin works. Anyway it's not a huge issue, from my side it works now with your trick, so no worries. But it could be useful to know for other people.

Also, thank you for your time and your active developement of this plugin.

Cheers

**Comment by axemwa** [comment]
No worries and sorry I'm tired and probably didn't explain it properly too I was on the phone as well.

The reason why this plugin requires you to ADD permissions (it doesn't actually negate them by default it just requires them to be added which is different to them being negated hence why you have to negate for Ops) is because by default it's meant to PREVENT players from stripping logs/wood which is why permissions are required if you want them to be able to use a command or axe, it is a permission based plugin hence why in the plugins subtitle it says "Prevent players from stripping logs and wood via **permission nodes** and **commands**." 

If I setup the plugin so that it allows players to strip logs by default it would be called something like StripControl that alone should be obvious enough as to why you have to negate permissions as an OP and add permissions (true) for ops and default players (who aren't opped) if you want them to allow something) because plugins that are designed to prevent interactions are supposed to be setup this way and most server owners would know this (not saying you're unintelligent or anything).

And the reason why I didn't write something like `For default players that are Ops you have to negate nostrip.* to default group` is because most server owners would already know this, it's common knowledge when the plugin is permission based. But regardless I am happy to add that message to the plugin page on your behalf considering you have helped me find some bugs and also for other people that might not have this knowledge as you didn't which is understandable, I guess for some it's not that obvious if you haven't used a permission based plugin that is designed to prevent interactions before.

So to summarize you probably just haven't come across a plugin that uses permissions like this before because you probably haven't used a plugin that is designed purely to prevent players from interacting with something and allowing them to use something with permissions but rather only commands or config settings which makes sense.

Where as chat plugins like VentureChat are NOT DESIGNED TO PREVENT CHAT BY DEFAULT but still give you the option to by negating them or via config/commands etc which is why that works differently if that makes sense so that is the only reason why plugins like that are setup differently permission wise.

Plus if you're using another permissions manager other than LuckPerms then that's another story again as I've heard of 1 other person having a permission related issue with PEX but no longer had it when they went to LuckPerms so there's that too.

Anyway I'll go ahead and add the `Fordefault players that are Ops you have to negate nostrip.* to default group (eg: LuckPerms Eg: /lp group default perm set worldedit.selection.* false). Then enable each permission to any group you want` message to my plugin page as you suggested for server owners that don't know about these types of prevention plugins. Thanks for your input it's given me a lot to think about for future releases so I can design it more for server owners who Op their players as well ;) Peace be with you.

**Comment by axemwa** [comment]
How's this, should be clear enough? Was a headache to put into words but I did it lol :)

EDITED

![image](https://user-images.githubusercontent.com/37608633/63226383-1f33c180-c218-11e9-949c-5e3de52ed225.png)

**Comment by JohOply** [comment]
First of all, I want to thank you for these huge explanations and the time you take. I just think the problem is really more simple than its appears.

To clarify, I don't want you to add ```For default players that are Ops you have to negate nostrip.* to default group``` because it's not my point. One more time, like I tried to explain to you, my point is **it seems have an issue with toggle command**. The fact is I experienced an issue (I explained in my first message on this issue). I done some tests, and your trick seemed works that's why I was focused on the permission topic. But it's not, because it wasn't the problem. I just tried again right now, and the issue still happens.

For some times, I thought there was a link between the issue and the fact player executed (or not) a right click before executing the command. But it's not. Now I think I have identified a process to reproduce the issue:
- By default, stripping is disable in config
- login to the server
- use /ns toggle --> it's enable and you are able to strip (test it)
- now /ns toggle again --> it's disable
- logout then login
- /ns toggle again --> it displays "enable", but it's not (test it)
- you need to toggle again to really enable stripping

If you join then leave then join again, issue appears too.

I tried with 2 different configurations:
- OP with all permissions
- Normal player with right permissions

In both cases, default group has "nostrip.*" permission set on false.

**Comment by axemwa** [comment]
That's ok I'm glad to help I just need more specific explanations like which player group(s) ur in when u test it eg: Admin group with op, VIP group with op, w/o op, are u remembering to leave admin group when testing in another group, does ur server deop you when you leave the server etc, cause no one else is having the problem and finding it very hard to reproduce especially considering that this plugin wasn't designed to work on players that have Op permissions it's purely permission based

What permissions exactly are u using for op and what permissions u using for normal player I'll see if I can reproduce the issue again

**Comment by JohOply** [comment]
Op/Admin: https://i.imgur.com/FNrprWs.png

Normal player with permissions: https://i.imgur.com/uA8pE3e.png

Normal player without permissions: only "nostrip.*" set to false

**Comment by axemwa** [comment]
Cause I've tried just about everything and none of us can reproduce the issue unless Opped without any permission and i wanna see if i can fix it for you.

**Comment by axemwa** [comment]
k thanks ill check it out

**Comment by axemwa** [comment]
both of those permissions are in default correct?

**Comment by JohOply** [comment]
Nop
Default has "nostrip.*" set to false.
The normal player is the player who has permissions to strip (rank 5), inherited from default

**Comment by axemwa** [comment]
so where are those permissions you sent me linked to? admin? or is it called rank 5?

**Comment by JohOply** [comment]
Admin is the first one
Player is the second one (called rank5)

**Comment by axemwa** [comment]
can u plz send me the "/lp editor" url

**Comment by axemwa** [comment]
when u do "/version nostrip" it definately says 2.0?

**Comment by JohOply** [comment]
> when u do "/version nostrip" it definately says 2.0?

Yes it is

**Comment by axemwa** [comment]
which group is toggling playing up on? as in which group are you in when you test this and is this when you are opped or not opped?

**Comment by axemwa** [comment]
?

**Comment by JohOply** [comment]
> which group is toggling playing up on? as in which group are you in when you test this and is this when you are opped or not opped?

Like I told you before

> I tried with 2 different configurations:
> 
> * OP with all permissions
> * Normal player with right permissions
> 
> In both cases, default group has "nostrip.*" permission set on false.

**Comment by axemwa** [comment]
yes but i dont know whether or not u keep changing it and the way ur stating it is very easy to misinterpret what ur trying to say so saying "like I said" and "like I told you before" doesn't actually help and need u to be clearer please thats why im asking u step by step now, and since ur group was called player i thought u were talkin about normal players in the default group lol so that made it hard to understand wat u meant too

**Comment by axemwa** [comment]
so by normal player do u mean the group player or a normal player as in the default group?

**Comment by JohOply** [comment]
OP with all permissions = admin group

normal player = player with permission to use the plugin

In my case/example, default group is useless because its has no permission so player can't use /ns toggle nor strip log

**Comment by axemwa** [comment]
ok now it makes sense thanks

**Comment by axemwa** [comment]
just making sure u werent added in both groups at the same time

**Comment by axemwa** [comment]
So just to clarify when u tested it in the admin group as an op the problem occurred. 

Then after that u removed ur self from the admin group and added yourself to the Player group to test on that and the problem still occurred when u were de opped?

I'm testing it now to see if i can reproduce

**Comment by JohOply** [comment]
Exactly

**Comment by axemwa** [comment]
In admin group you're supposed to use `.*` not just `*` doing that fixes it
I'll check the Player group next

**Comment by axemwa** [comment]
Player group works fine for me deopped

**Comment by JohOply** [comment]
You experience issue only when you're opped?

**Comment by axemwa** [comment]
yes opped in "default" group if I don't use negate permissions but I can fix that

**Comment by axemwa** [comment]
or u mean player? Ill test in player group opped havent done that yet only deopped

**Comment by JohOply** [comment]
I mean, you experience issue only when you're /op AND in admin group 
or /op AND in player group? (I still speak about player with right permissions to use the plugin)

**Comment by axemwa** [comment]
oh yeah it happens for me managed to reproduce issue in Player I'll see if I can do an Op check for that, here's a test version see if it fixes it https://www.dropbox.com/s/1if6d7wbwt1qc7d/NoStrip-2.0.jar?dl=1

**Comment by axemwa** [comment]
Couldn't reproduce issue in admin after i changed `*` to `.*` so that was the problem for that group you just forgot to add the preiod `.` before the asterisk `*` so its supposed to be this `.*` not this `*` in ur admin permissions.

Also Player group works fine for me I retested after removing myself from "admin" group and adding myself to the "Player" group and making sure I was still in the "default" group as well as checking for opped and deopped and all seems to work fine using latest release version on spigot. And be sure that when adding/removing yourself from a group to do it in game or via console and not via the /lp editor because it can sometimes think you're adding/removing a permission not a group

PS: I removed the test jar file from dropbox that I uploaded here on the thread cause it broke toggle alltogether for normal players so dont use it anymore if u happened to have seen it and tested it

EDIT - Ok just did a tonne of testing and I can reproduce the issue indefinitely I finally get what u mean now it was just the way you were explaining it that was confusing me.

So what it's doing is when `nostrip.toggle` and `nostrip.axe.*` permissions are given to "default" group (and are the only permissions given opped or not) is when I join and use `/ns toggle` to enable stripping and then do the command again to disable it then leave the server and rejoin, if I use `/ns toggle` to enable it, it doesn't actually enable stripping even though it says it does unless I do the command once again, that's what u were trying to explain to me correct? 

I humbly apologize it's not that I didn't believe you it was just the way you were explaining how to reproduce the issue that was confusing me and made me think you were talking about something else but now I know exactly what the problem is and how to reproduce it and I should be able to fix it in no time, I just gotta find the time to do it. Thanks for reporting it

**Comment by axemwa** [comment]
Ok I have fixed the issue with the enabled/disabled toggles not working properly when you leave and rejoin the server here's the fixed version, I've tested several times and it seems to have fixed the issue thanks so much for pointing it out, it took me longer than expected but it seems to be fixed. 

I just had to separate the checks in the code as some were merged together which was causing that issue. Hopefully the fix didn't create an additional issue but it seems to be working well.

https://www.spigotmc.org/resources/nostrip.61574/download?version=290220
